Digital Credentials: implement Digital Credentials Coordinator and IPC

marcoscaceres opened the pull request:
#### f4419f04902e3be11bcc20229648c197968702f4
<pre>
Digital Credentials: implement Digital Credentials Coordinator and IPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=277850">https://bugs.webkit.org/show_bug.cgi?id=277850</a>
<a href="https://rdar.apple.com/134026261">rdar://134026261</a>

Reviewed by Jean-Yves Avenard.

The patch introduces a Digital Credentials Coordinator that manages digital credential requests and integrates a new picker UI component (`WKDigitalCredentialsPicker`) to allow user interaction.
This is part of the implementation for the Digital Credentials API, which provides a way for web applications to request and use digital credentials from the user&apos;s device. Spec:
<a href="https://wicg.github.io/digital-credentials/">https://wicg.github.io/digital-credentials/</a>

At at high level, this patch:

- Integrates digital credentials support into WebKit&apos;s existing credentials management infrastructure.
- Ensures digital credentials requests are handled asynchronously across process boundaries with robust error handling.
- Provides a friendly interface for selecting and returning digital credentials on supported platforms.

Key Changes:

- Digital Credentials Coordinator
  - Receives digital credential requests and validates preconditions (e.g. document focus, visibility, and permissions), as per spec.
  - Delegates the presentation of a picker UI to the platform-specific UI layer.
  - Handles completion callbacks by resolving or rejecting promises based on user input or errors.

- New UI Component - WKDigitalCredentialsPicker
  - Implements the picker UI in Objective-C++.
  - Provides methods to present and dismiss the picker.
  - Returns the digital credentials response data (or an error) via an asynchronous completion handler.

- IPC and Asynchronous Messaging
  - Adds new IPC messages (`ShowDigitalCredentialsPicker` and `DismissDigitalCredentialsPicker`) to route requests between the WebProcess and UIProcess.
  - Updates related classes (e.g. WebPage, PageClient, and their platform-specific implementations) to integrate digital credentials actions.

- Preferences and Feature Flags
  - Introduces a new preference key (`WebKitDigitalCredentialsEnabledPreferenceKey`) to enable or disable the Digital Credentials API.

- Platform Integration and Refactoring
  - Removes legacy digital credentials proxy code, consolidating functionality into the new coordinator.
  - Updates iOS and macOS clients (e.g. `PageClientImplIOS`, `PageClientImplMac`, and `WKContentViewInteraction`) so they can present and dismiss the digital credentials picker.
  - Provides a dummy client implementation for unsupported platforms to ensure graceful degradation.

* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/disabled-by-permissions-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/enabled-on-self-origin-by-permissions-policy.https.sub-expected.txt:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::get):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
(WebCore::CredentialRequestCoordinator::create):
(WebCore::CredentialRequestCoordinator::CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::PickerStateGuard::PickerStateGuard):
(WebCore::CredentialRequestCoordinator::PickerStateGuard::~PickerStateGuard):
(WebCore::CredentialRequestCoordinator::currentState const):
(WebCore::CredentialRequestCoordinator::canTransitionTo const):
(WebCore::CredentialRequestCoordinator::setState):
(WebCore::CredentialRequestCoordinator::setCurrentPromise):
(WebCore::CredentialRequestCoordinator::currentPromise):
(WebCore::CredentialRequestCoordinator::presentPicker):
(WebCore::CredentialRequestCoordinator::handleDigitalCredentialsPickerResult):
(WebCore::CredentialRequestCoordinator::parseDigitalCredentialsResponseData const):
(WebCore::CredentialRequestCoordinator::finalizeDigitalCredential):
(WebCore::CredentialRequestCoordinator::abortPicker):
(WebCore::CredentialRequestCoordinator::~CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::discoverFromExternalSource): Deleted.
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::DigitalCredential::discoverFromExternalSource):
* Source/WebCore/Modules/identity/DigitalCredential.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequestOptions.h:
* Source/WebCore/Modules/identity/DigitalCredentialsRequestData.h: Copied from Source/WebCore/Modules/identity/DigitalCredentialRequest.h.
* Source/WebCore/Modules/identity/DigitalCredentialsResponseData.h: Copied from Source/WebCore/Modules/identity/DigitalCredentialRequest.h.
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp: Removed.
* Source/WebCore/Modules/identity/dummy/DummyCredentialRequestCoordinatorClient.cpp: Renamed from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
(WebCore::DummyCredentialRequestCoordinatorClient::showDigitalCredentialsPicker):
(WebCore::DummyCredentialRequestCoordinatorClient::dismissDigitalCredentialsPicker):
* Source/WebCore/Modules/identity/dummy/DummyCredentialRequestCoordinatorClient.h: Copied from Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::showDigitalCredentialsPicker):
(WebCore::Chrome::dismissDigitalCredentialsPicker):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::showDigitalCredentialsPicker):
(WebCore::ChromeClient::dismissDigitalCredentialsPicker):
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/Page.h:
(WebCore::Page::credentialRequestCoordinator):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/Logging.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showDigitalCredentialsPicker:completionHandler:]):
(-[WKWebView _dismissDigitalCredentialsPicker:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.messages.in: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.h: Copied from Source/WebCore/Modules/identity/CredentialRequestCoordinator.h.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm: Added.
(-[WKDigitalCredentialsPickerDelegate initWithDigitalCredentialsPickerDelegate:]):
(-[WKRequestDataResult initWithRequestDataBytes:protocol:]):
(-[WKDigitalCredentialsPicker initWithView:]):
(-[WKDigitalCredentialsPicker delegate]):
(-[WKDigitalCredentialsPicker setDelegate:]):
(-[WKDigitalCredentialsPicker presentWithRequestData:completionHandler:]):
(-[WKDigitalCredentialsPicker dismissWithCompletionHandler:]):
(-[WKDigitalCredentialsPicker dismiss]):
(-[WKDigitalCredentialsPicker dismissWithResponse:]):
(-[WKDigitalCredentialsPicker completeWith:]):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::showDigitalCredentialsPicker):
(WebKit::PageClient::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::showDigitalCredentialsPicker):
(WebKit::WebPageProxy::dismissDigitalCredentialsPicker):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::showDigitalCredentialsPicker):
(WebKit::PageClientImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shareSheetDidDismiss:]):
(-[WKContentView _showDigitalCredentialsPicker:completionHandler:]):
(-[WKContentView _dismissDigitalCredentialsPicker:]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::showDigitalCredentialsPicker):
(WebKit::PageClientImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showDigitalCredentialsPicker):
(WebKit::WebViewImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
(WebKit::DigitalCredentialsCoordinator::showDigitalCredentialsPicker):
(WebKit::DigitalCredentialsCoordinator::dismissDigitalCredentialsPicker):
(WebKit::DigitalCredentialsCoordinator::requestDigitalCredential): Deleted.
(WebKit::DigitalCredentialsCoordinator::cancel): Deleted.
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
(): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::showDigitalCredentialsPicker):
(WebKit::WebChromeClient::dismissDigitalCredentialsPicker):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::showDigitalCredentialsPicker):
(WebKit::WebPage::dismissDigitalCredentialsPicker):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/WebView/WebPreferenceKeysPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(-[WebPreferences digitalCredentialsEnabled]):
(-[WebPreferences setDigitalCredentialsEnabled:]):
* Source/WebKitLegacy/mac/WebView/WebPreferencesPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: <a href="https://commits.webkit.org/290115@main">https://commits.webkit.org/290115@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7018d9d2610c3576b3a711baea542ce44c681e73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88999 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43718 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93971 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16707 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68568 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26245 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6801 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80682 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48931 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6550 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38867 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81798 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36033 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95810 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87775 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16179 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11822 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77445 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16435 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76491 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76734 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18916 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21127 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9262 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16193 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21504 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110268 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15934 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26464 "Found 16 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.default, microbenchmarks/memcpy-wasm-medium.js.mini-mode, microbenchmarks/memcpy-wasm-medium.js.no-llint, microbenchmarks/memcpy-wasm-small.js.default, microbenchmarks/set-delete-add.js.no-llint, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-collect-continuously, wasm.yaml/wasm/regress/llint-callee-saves-without-fast-memory.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19385 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17715 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebCore/Modules/identity/DigitalCredentialsRequestData.h:
```diff
@@ -0,0 +1,43 @@
+
+struct DigitalCredentialsRequestData {
+    DigitalCredentialRequestOptions options;
+    String url;
```
URL type rather than a String?

marcoscaceres replied:
I wonder if we should use a `WebCore::SecurityOriginData` and call this topOrigin instead... `url` is not super clear.

sheeparegreat commented on Source/WebCore/Modules/identity/DigitalCredentialsResponseData.h:
```diff
@@ -0,0 +1,41 @@
+
+struct DigitalCredentialsResponseData {
+    IdentityCredentialProtocol protocol;
+    String responseData;
```
Can this be broken down further?

marcoscaceres replied:
Possibly, yes... we don't yet have a response format.

marcoscaceres replied:
After some investigation, turns out that it can't. It's an encrypted blob, basically.

sheeparegreat commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -6398,22 +6398,36 @@ struct WebCore::MockWebAuthenticationConfiguration {
-struct WebCore::OpenID4VPRequest {
+#if HAVE(DIGITAL_CREDENTIALS_UI)
+
+[Nested] struct WebCore::OpenID4VPRequest {
```
Is this currently a placeholder? What are the members here?

marcoscaceres replied:
yep. Just a placeholder for now.

jyavenard commented on LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute.https-expected.txt:
```diff
@@ -1,9 +1,9 @@
 
 
 PASS Policy to use: null, is cross-origin: false, is allowed by policy: true
-FAIL Policy to use: null, is cross-origin: true, is allowed by policy: false assert_equals: <iframe src="https://127.0.0.1:9443/digital-credentials/support/iframe.html" data-expect-is-allowed="false"></iframe> - Not implemented expected "NotAllowedError" but got "NotSupportedError"
```
typically don't expect to see assertions showing in the test results !

marcoscaceres replied:
yeah, these are hard to debug without them. Why we left them in there.

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -28,100 +28,273 @@
+
+CredentialRequestCoordinator::CredentialRequestCoordinator(UniqueRef<CredentialRequestCoordinatorClient>&& client, Page& page)
     : m_client(WTFMove(client))
+    , m_page(WeakPtr { page })
```
```suggested
    , m_page(page)
```

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -28,100 +28,273 @@
+}
+
+CredentialRequestCoordinator::PickerStateGuard::PickerStateGuard(PickerStateGuard&& other) noexcept
+    : m_coordinator(other.m_coordinator)
```
can man use move semantics there too

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -28,100 +28,273 @@
+    auto weakThis = WeakPtr { *this };
+    m_client->showDigitalCredentialsPicker(
+        requestData,
+        [weakThis, signal = RefPtr { signal }](Expected<DigitalCredentialsResponseData, ExceptionData>&& responseOrException) {
```
```suggested
        [weakThis = WeakPtr { *this }, signal](Expected<DigitalCredentialsResponseData, ExceptionData>&& responseOrException) {
```

no need for intermediary weakThis variable. signal can be copy captured already, it's a RefPtr, no need for `RefPtr { ... }` again

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -28,100 +28,273 @@
+    m_client->showDigitalCredentialsPicker(
+        requestData,
+        [weakThis, signal = RefPtr { signal }](Expected<DigitalCredentialsResponseData, ExceptionData>&& responseOrException) {
+            if (!weakThis)
```
```suggested
if (RefPtr protectedThis = weakThis.get())
```

to follow no code hardening rules

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -28,100 +28,273 @@
+        return;
+    }
+
+    auto credential = DigitalCredential::create(
```
`   Ref credential = DigitalCredential::create(`

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -27,38 +27,76 @@
+        PickerStateGuard(PickerStateGuard&&) noexcept;
+        PickerStateGuard& operator=(PickerStateGuard&&) noexcept;
+
+        ~PickerStateGuard();
```
either make destructor virtual, or make PickerStateGuard `final`

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -27,38 +27,76 @@
+        Idle,
+        Presenting,
+        Aborting
+    }; // enum class PickerState
```
unnecessary comment

marcoscaceres replied:
I want to keep these comments... it's a bit confusing without it in a code editor I found.

jyavenard commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -27,38 +27,76 @@
-    bool m_isCancelling = false;
-    CompletionHandler<void()> m_queuedRequest;
+    explicit CredentialRequestCoordinator(UniqueRef<CredentialRequestCoordinatorClient>&&, Page&);
+    UniqueRef<CredentialRequestCoordinatorClient> m_client;
```
m_client should be const

jyavenard commented on Source/WebCore/Modules/identity/DigitalCredential.cpp:
```diff
@@ -47,6 +63,60 @@ DigitalCredential::DigitalCredential(JSC::Strong<JSC::JSObject>&& data, Identity
+    requestData.topOrigin = document.protectedTopOrigin()->data().isolatedCopy();
+    requestData.documentOrigin = document.protectedSecurityOrigin()->data().isolatedCopy();
+
+    Ref coordinator = document.frame()->page()->credentialRequestCoordinator();
```
shouldn't you test that frame() or page() aren't null?

marcoscaceres replied:
Tested them on line 80, but will switch to protectedFrame()->protectedPage()

jyavenard commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -51,27 +49,20 @@ RefPtr<WebPage> DigitalCredentialsCoordinator::protectedPage() const
-        return;
-    }
-    page->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::RequestDigitalCredential(webFrame->frameID(), webFrame->info(), options), WTFMove(handler));
+    if (auto page = protectedPage())
```
```suggested
    if (RefPtr page = protectedPage())
```
do not use `auto` if the type is Ref, RefPtr or RetainPtr (new code hardening style)

jyavenard commented on Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
```diff
@@ -51,27 +49,20 @@ RefPtr<WebPage> DigitalCredentialsCoordinator::protectedPage() const
-    }
-
-    page->sendWithAsyncReply(Messages::DigitalCredentialsCoordinatorProxy::Cancel(), WTFMove(handler));
+    if (auto page = protectedPage())
```
ditto

jyavenard reviewed (approved):
LGTM.

r+ with comments addressed.

philn commented on Source/WebKit/Sources.txt:
```diff
@@ -563,7 +563,7 @@ UIProcess/Automation/WebAutomationSession.cpp @no-unify
 UIProcess/Downloads/DownloadProxyMap.cpp
 
-UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp
+UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm
```
the mm file should likely be added to SourcesCocoa.txt.

marcoscaceres committed:
Digital Credentials: implement Digital Credentials Coordinator and IPC
https://bugs.webkit.org/show_bug.cgi?id=277850
rdar://134026261

Reviewed by Jean-Yves Avenard.

The patch introduces a Digital Credentials Coordinator that manages digital credential requests and integrates a new picker UI component (`WKDigitalCredentialsPicker`) to allow user interaction.
This is part of the implementation for the Digital Credentials API, which provides a way for web applications to request and use digital credentials from the user's device. Spec:
https://wicg.github.io/digital-credentials/

At at high level, this patch:

- Integrates digital credentials support into WebKit's existing credentials management infrastructure.
- Ensures digital credentials requests are handled asynchronously across process boundaries with robust error handling.
- Provides a friendly interface for selecting and returning digital credentials on supported platforms.

Key Changes:

- Digital Credentials Coordinator
  - Receives digital credential requests and validates preconditions (e.g. document focus, visibility, and permissions), as per spec.
  - Delegates the presentation of a picker UI to the platform-specific UI layer.
  - Handles completion callbacks by resolving or rejecting promises based on user input or errors.

- New UI Component - WKDigitalCredentialsPicker
  - Implements the picker UI in Objective-C++.
  - Provides methods to present and dismiss the picker.
  - Returns the digital credentials response data (or an error) via an asynchronous completion handler.

- IPC and Asynchronous Messaging
  - Adds new IPC messages (`ShowDigitalCredentialsPicker` and `DismissDigitalCredentialsPicker`) to route requests between the WebProcess and UIProcess.
  - Updates related classes (e.g. WebPage, PageClient, and their platform-specific implementations) to integrate digital credentials actions.

- Preferences and Feature Flags
  - Introduces a new preference key (`WebKitDigitalCredentialsEnabledPreferenceKey`) to enable or disable the Digital Credentials API.

- Platform Integration and Refactoring
  - Removes legacy digital credentials proxy code, consolidating functionality into the new coordinator.
  - Updates iOS and macOS clients (e.g. `PageClientImplIOS`, `PageClientImplMac`, and `WKContentViewInteraction`) so they can present and dismiss the digital credentials picker.
  - Provides a dummy client implementation for unsupported platforms to ensure graceful degradation.

* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/disabled-by-permissions-policy.https.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/enabled-on-self-origin-by-permissions-policy.https.sub-expected.txt:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
(WebCore::CredentialsContainer::get):
* Source/WebCore/Modules/credentialmanagement/CredentialsContainer.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
(WebCore::CredentialRequestCoordinator::create):
(WebCore::CredentialRequestCoordinator::CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::PickerStateGuard::PickerStateGuard):
(WebCore::CredentialRequestCoordinator::PickerStateGuard::~PickerStateGuard):
(WebCore::CredentialRequestCoordinator::currentState const):
(WebCore::CredentialRequestCoordinator::canTransitionTo const):
(WebCore::CredentialRequestCoordinator::setState):
(WebCore::CredentialRequestCoordinator::setCurrentPromise):
(WebCore::CredentialRequestCoordinator::currentPromise):
(WebCore::CredentialRequestCoordinator::presentPicker):
(WebCore::CredentialRequestCoordinator::handleDigitalCredentialsPickerResult):
(WebCore::CredentialRequestCoordinator::parseDigitalCredentialsResponseData const):
(WebCore::CredentialRequestCoordinator::finalizeDigitalCredential):
(WebCore::CredentialRequestCoordinator::abortPicker):
(WebCore::CredentialRequestCoordinator::~CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::discoverFromExternalSource): Deleted.
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::DigitalCredential::discoverFromExternalSource):
* Source/WebCore/Modules/identity/DigitalCredential.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequestOptions.h:
* Source/WebCore/Modules/identity/DigitalCredentialsRequestData.h: Copied from Source/WebCore/Modules/identity/DigitalCredentialRequest.h.
* Source/WebCore/Modules/identity/DigitalCredentialsResponseData.h: Copied from Source/WebCore/Modules/identity/DigitalCredentialRequest.h.
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp: Removed.
* Source/WebCore/Modules/identity/dummy/DummyCredentialRequestCoordinatorClient.cpp: Renamed from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
(WebCore::DummyCredentialRequestCoordinatorClient::showDigitalCredentialsPicker):
(WebCore::DummyCredentialRequestCoordinatorClient::dismissDigitalCredentialsPicker):
* Source/WebCore/Modules/identity/dummy/DummyCredentialRequestCoordinatorClient.h: Copied from Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::showDigitalCredentialsPicker):
(WebCore::Chrome::dismissDigitalCredentialsPicker):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::showDigitalCredentialsPicker):
(WebCore::ChromeClient::dismissDigitalCredentialsPicker):
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/Page.h:
(WebCore::Page::credentialRequestCoordinator):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/Logging.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWebPushDaemonConnection.cpp:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _showDigitalCredentialsPicker:completionHandler:]):
(-[WKWebView _dismissDigitalCredentialsPicker:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.cpp: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.h: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/DigitalCredentialsCoordinatorProxy.messages.in: Removed.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.h: Copied from Source/WebCore/Modules/identity/CredentialRequestCoordinator.h.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm: Added.
(-[WKDigitalCredentialsPickerDelegate initWithDigitalCredentialsPickerDelegate:]):
(-[WKRequestDataResult initWithRequestDataBytes:protocol:]):
(-[WKDigitalCredentialsPicker initWithView:]):
(-[WKDigitalCredentialsPicker delegate]):
(-[WKDigitalCredentialsPicker setDelegate:]):
(-[WKDigitalCredentialsPicker presentWithRequestData:completionHandler:]):
(-[WKDigitalCredentialsPicker dismissWithCompletionHandler:]):
(-[WKDigitalCredentialsPicker dismiss]):
(-[WKDigitalCredentialsPicker dismissWithResponse:]):
(-[WKDigitalCredentialsPicker completeWith:]):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::showDigitalCredentialsPicker):
(WebKit::PageClient::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::didDestroyFrame):
(WebKit::WebPageProxy::didStartProvisionalLoadForFrameShared):
(WebKit::WebPageProxy::showDigitalCredentialsPicker):
(WebKit::WebPageProxy::dismissDigitalCredentialsPicker):
(WebKit::WebPageProxy::resetState):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::showDigitalCredentialsPicker):
(WebKit::PageClientImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shareSheetDidDismiss:]):
(-[WKContentView _showDigitalCredentialsPicker:completionHandler:]):
(-[WKContentView _dismissDigitalCredentialsPicker:]):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::showDigitalCredentialsPicker):
(WebKit::PageClientImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showDigitalCredentialsPicker):
(WebKit::WebViewImpl::dismissDigitalCredentialsPicker):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.cpp:
(WebKit::DigitalCredentialsCoordinator::showDigitalCredentialsPicker):
(WebKit::DigitalCredentialsCoordinator::dismissDigitalCredentialsPicker):
(WebKit::DigitalCredentialsCoordinator::requestDigitalCredential): Deleted.
(WebKit::DigitalCredentialsCoordinator::cancel): Deleted.
* Source/WebKit/WebProcess/DigitalCredentials/DigitalCredentialsCoordinator.h:
(): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::showDigitalCredentialsPicker):
(WebKit::WebChromeClient::dismissDigitalCredentialsPicker):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::showDigitalCredentialsPicker):
(WebKit::WebPage::dismissDigitalCredentialsPicker):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKitLegacy/mac/WebView/WebPreferenceKeysPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(-[WebPreferences digitalCredentialsEnabled]):
(-[WebPreferences setDigitalCredentialsEnabled:]):
* Source/WebKitLegacy/mac/WebView/WebPreferencesPrivate.h:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: https://commits.webkit.org/290115@main
